rework tenant attach code to share the initialization code path with tenant load

problame opened the pull request:
At the time of writing, the only logical difference between Attach and Load is that Attach learns the list of timelines by querying the remote storage, whereas Load learns it by listing the timelines directory of the tenant.

This patch restructures the code such that Attach

1. prepares the on-disk state and then
2. calls into the same `load()` routine that is used by Load.

Further, this patch provides the following fixes & improvements to Attach:

1. Make Attach durable before acknowledging it to the management API client.

   Before this change, we would acknowledge after just creating the in-memory tenant.
   In the event of a crash before creating the tenant directory & fsyncing the attaching marker,
   the pageserver would come up without the tenant present (404), even though we acknowledged
   success to the client.

2. Simplified resume logic if we crash during Attach.
   Before this patch, if we crashed during Attach with some timelines downloaded and others not downloaded,
   we would combine existing metadata files with remote ones one-by-one to figure out what's missing.
   That was necessary before on-demand download because we were downloading layer files as part of Attach.
   However, with on-demand download, Attach only downloads & writes the timeline metadata files.

   After this patch, when we crash during Attach, we blow away the tenant's directory while leaving its attach marker file in place. Then, we start over.
   
   IMO this is significantly easier to reason about compared to what we had before.
   Note that we were losing the work for the downloads even before this change, so that's not a regression (the old reconcile_with_remote would still need to download the `IndexPart`s when resuming Attach after a crash).

   If we want to improve on this in the future, I think the first order of business will be to
   avoiding re-downloading the `IndexPart`'s and the initial `list_remote_timelines()`.
   However, given that crashes should be rare, and attach events also, I don't think the number
   one priority with Attach code should be to make it as simple as possible.

For (2), I changed the location of the attach marker file to be outside the tenant directory,
so that we can use standard functions for removing the tenant directory.
I even wrote a migration function for it, although in retrospect, I think it's quite unlikely that there are any tenants in attaching state deployed.
But oh well, now the code is there, and it even has unit tests.
We can delete the migration code once we've successfully rolled it out to all regions.

The remaining wrinkle with this change is that Attach needs to hint the downloaded `IndexPart`s to `load()` so that it doesn't download them twice during Attach, which would be wasteful.
The mechanism for this is the new `TenantLoadReason` and `TimelineLoadReason`.

We could eliminate this particular case by on-demand downloading the metadata.
However, that might open up another can of worms which I'd like to avoid.
If we ever want to go that route, I suggest we start tracking the attachment state of a timeline more formally, e.g., in a `timelines.json` file.

---

TODOs:

- [x] reviews
- [x] squash & update commit message

koivunej commented:
Related just merged (before you continue work): #3446

problame commented:
Rebased to get fixes for GitHub workflows from `main`.

koivunej commented on pageserver/src/config.rs:
```diff
@@ -459,8 +459,8 @@ impl PageServerConf {
     pub fn tenant_attaching_mark_file_path(&self, tenant_id: &TenantId) -> PathBuf {
-        self.tenant_path(tenant_id)
-            .join(TENANT_ATTACHING_MARKER_FILENAME)
+        // TODO deal with backward compatiblity by renaming old attaching marker file
```
Stale TODO, isn't this already implemented?

koivunej commented on pageserver/src/tenant.rs:
```diff
@@ -570,12 +705,66 @@ impl Tenant {
+        }
+
+        // create the marker file, i.e., /tenants/${tenant_id}___attaching
+        fs::File::create(&marker_file).map_err(|error| AttachError::CreateMarkerFile {
```
```suggestion
        std::fs::File::create(&marker_file).map_err(|error| AttachError::CreateMarkerFile {
```

Probably better to move these as either `std` or `tokio` since we now have so much of both on the same path even.

koivunej commented on pageserver/src/tenant.rs:
```diff
@@ -1036,40 +1215,52 @@ impl Tenant {
+                    remote_startup_data,
+                    remote_client,
+                } = *boxed;
+                info!("re-using index_part download earlier by attach");
```
A bit skeptical if this should be logged at all.

koivunej commented on pageserver/src/tenant.rs:
```diff
@@ -997,16 +1145,46 @@ impl Tenant {
+            TenantLoadReason::Attaching {
+                mut attach_data_by_timeline_id,
+            } => {
+                let mut merged = HashMap::new();
```
```suggestion
                let mut merged = HashMap::with_capacity(timelines_to_load.len());
```

Alternative is to turn the `for (_, _) in timelines.to_load.drain()` to a iterator like which also carries the size hint used for `with_capacity`:

```rust
timelines_to_load.drain()
  .map(|(id, meta)| {
    let val = attach_data_by_timeline_id.remove(&id).context("...")?;
    let val = TimelineLoadReason::Attach(data);
    Ok((id, (meta, val)))
  })
  .collect::<Result<HashMap<_, _>, _>()?;
```

(The `FromIterator` impl used here is better explained in doc comment: https://doc.rust-lang.org/stable/src/core/result.rs.html#2030-2077)

This would actually make it more clear that:
- `assert!(replaced.is_none())` -- the two maps `merged` and `timelines_to_laod` both have the same key, impossible for this to happen

koivunej reviewed (commented):
All in all good changes, smaller nits, rename ideas from me so far. I probably missed some important points so will probably come back to this. Would definitetly like to see the added test cases split at least.

koivunej commented:
(Re-ran failed steps in an effort to help the work with runners.)

problame commented on pageserver/src/tenant.rs:
```diff
@@ -109,7 +112,41 @@ pub use crate::tenant::timeline::WalReceiverInfo;
+}
+
+enum TenantLoadReason {
+    Regular,
```
The `/load` endpoint also uses it. Have a third enum variant for it, just because we can?

problame replied:
Pushed commit `differentiated TenantLoadReason::NotAttaching, including differentia...` to address this. Not exactly pretty, but explicit and properly differentiated, so that future readers of the code understand immediately what code paths share the tenant-load code.

koivunej replied:
Aah I see `/load` is a counterpart of `/ignore`? I only remember discussion about adding an `/reload` but I guess this is it.

[07357ef](https://github.com/neondatabase/neon/pull/3466/commits/07357ef0b0e10a52fdd95e66e932151711cc86a2) is the commit.

NotAttach vs. Attach makes me think if an Option would be applicable but perhaps this at least makes one think what other kinds of calls there might be than attach.

problame commented on pageserver/src/tenant.rs:
```diff
@@ -694,56 +901,40 @@ impl Tenant {
+                    assert!(replaced.is_none());
+                }
+                Err(e) => {
+                    // XXX print timeline id here, would be easy with JoinMap, but it's not stable yet.
```
Huh, now that I think about it, the error message already includes 
```
                .map(move |res| {
                    res.with_context(|| format!("download index part for timeline {timeline_id}"))
                })
```

So I think the right `error!` message here is

```
                    error!("failed to {e:#}");
```

as per https://neondb.slack.com/archives/C0277TKAJCA/p1673453329770429

koivunej commented on pageserver/src/tenant.rs:
```diff
@@ -552,14 +596,120 @@ impl Tenant {
+        loop {
+            let dir_entry = match dir_entries.next() {
+                None => break,
+                Some(res) => res.context("reat tenants dir entry")?,
```
```suggestion
                Some(res) => res.context("read tenants dir entry")?,
```

koivunej commented on pageserver/src/tenant.rs:
```diff
@@ -552,14 +596,120 @@ impl Tenant {
+
+            let legacy_marker_parent = legacy_attach_marker_path.parent().unwrap();
+
+            match legacy_marker_parent
+                .file_name()
+                .and_then(OsStr::to_str)
+                .unwrap()
+                .parse::<TenantId>()
```
This unwrap panicking means that we have a non-utf8 path amongst our tenant dirs. I guess I'd prefer to have expect but unsure about the message I'd suggest.

Since this feels so repeated, I think it'd be important to just parse OsStr to an Id but because of windows that is not so easy either.

Well, just noting this. I guess a panic hit here would be easy to resolve by moving the offending dir out. In general it'd be nice to write this as:

```suggestion
            match legacy_marker_parent
                .file_name()
                .and_then(OsStr::to_str)
                .map(|s| s.parse::<TenantId>().map_err(anyhow::Error::new))
                .ok_or_else(|| anyhow::anyhow!("non utf8 path"))
```

And match instead on `Some(was_utf8_convertable_path)` or it wasn't `None`.

problame replied:
So, this was basically copy-pasta from `fn init_tenant_mgr`.

And there's a similar pattern in `Tenant::load` for parsing timeline dir names.

Created https://github.com/neondatabase/neon/issues/3488

koivunej replied:
Sounds good.

problame commented on pageserver/src/tenant.rs:
```diff
@@ -109,7 +112,55 @@ pub use crate::tenant::timeline::WalReceiverInfo;
+pub const TENANT_ATTACHING_LEGACY_MARKER_FILENAME: &str = "attaching";
+pub const TENANT_ATTACHING_MARKER_SUFFIX: &str = "___attaching";
+
+/// The error message does not include the tenant ID.
```
IMO a function should not repeat its arguments in the errors it emits.
Rationale:
- avoids unnecessary copies if the caller has distinguished error handling (i.e., a `match`, not simply `?`)
- DRY: with your proposal, each error variant needs to be enrichted with `tenant_id`. With my approach, the caller is responsible for enrichting the error, in exactly one place.

SomeoneToIgnore replied:
You could've wrapped the enum with some struct or enforced the ID presence some other way, but up to you.

problame commented on pageserver/src/tenant/mgr.rs:
```diff
@@ -421,13 +443,8 @@ pub async fn attach_tenant(
-
-        let tenant = Tenant::spawn_attach(conf, tenant_id, remote_storage, ctx);
+        let tenant = Tenant::spawn_start_attach(conf, tenant_id, remote_storage, ctx)
+            .map_err(|source| AttachError { tenant_id, source })?;
```
> AttachError from above is used only here, and that gets ?-ed into TenantMapInsertError, via more explicit route than anyhow::Error before.
> Am I wrong, or all related changes could be replaced with a match and anyhow::anyhow!(attach tenant {tenant_id}: {source:#}) returned one way or another?
>  Looks simpler the latter way to me, unless it's a preparation for something bigger here.

Yeah, right, there is no distinguished treatment.
Will change that.

>  We can add another variant into TenantMapInsertError.

Nah, if we ever need distinguished treatment of attach errors in the caller,  I'd rather make it's `Closure` variant generic, i.e., `enum TenantMapInsertError<OE> { ...,  Other(OE) }`.

problame commented:
- Addressed review comment https://github.com/neondatabase/neon/pull/3466#discussion_r1091626830 
- Squashed commits & updated commit message to write-up in the PR description above.

LizardWizzard reviewed (commented):
Thanks for cleaning this up! I like that it became simpler. Left some comments, feel free to just resolve NIT ones.  Found nothing serious. If you'd ask I'm +1 for not handling attach marker migration at all. We just need to be sure no one is running attaches during release

problame commented:
Thanks for the review, will address the nits tomorrow.

> If you'd ask I'm +1 for not handling attach marker migration at all. We just need to be sure no one is running attaches during release

You think the risk of a bug in the migration code outweighs the risk that there might be some attach markers somewhere?

Idea: 
- remove the migration code
- make this PR modify the deploy Ansible script to check that there's no old attach markers present before deploying the new binary. 
- after the deployment, remove that check from the deploy Ansible script

problame commented on pageserver/src/tenant.rs:
```diff
@@ -381,18 +432,13 @@ impl Drop for TimelineUninitMark {
 //
 // So the solution is to take remote metadata only when we're attaching.
```
Please review https://github.com/neondatabase/neon/pull/3517 and resolve once merged.

problame commented on pageserver/src/tenant.rs:
```diff
@@ -612,64 +826,62 @@ impl Tenant {
-            crashsafe::fsync_file_and_parent(&marker_file)
-                .context("fsync tenant attaching marker file and parent")?;
+            info!("removing prior attach operation's progress");
+            std::fs::remove_dir_all(&tenant_dir).context("remove attaching tenant dir")?;
```
> . We can write the config to attach marker. Or do it in some other way. WDYT?

Yeah, that's how I'd solve it in the future. Should we prepare for this by writing
```
{ "format_version": ` }
```
into the new attach markers?

LizardWizzard replied:
For now its simple, the file is empty, so we can check for it and think about format later

problame replied:
The future version won't be able to distinguish a partially written file that it wrote before a crash from a file written by a version with this patch, then.

I'll add this now, this PR is in no big rush anyways, let's get it right now.

problame replied:
How does https://github.com/neondatabase/neon/pull/3519 look like to you?

LizardWizzard replied:
Reviewed #3519

LizardWizzard commented on pageserver/src/tenant.rs:
```diff
@@ -570,12 +720,66 @@ impl Tenant {
+        }
+
+        // create the marker file, i.e., /tenants/${tenant_id}___attaching
+        std::fs::File::create(&marker_file).map_err(|error| AttachError::CreateMarkerFile {
```
There can be a race between exists check and actual call to create. 

> I suppose we shold be using create_new(true) here?

Right! I didnt realize that usual create will truncate the file if it already exists instead of returning an error

LizardWizzard commented:
> You think the risk of a bug in the migration code outweighs the risk that there might be some attach markers somewhere?

Realistically I think there wont be any attach markers, and its easy to check for that before the release. So it wont trigger any possible bugs in migration code.

So IMO this PR illustrates the need for automated changelog entry generation. I e there should be a possibility to leave a note that will be picked up in release PR so person who will deploy a release can check for additional safety requirements. cc @vadim2404 @shanyp 

In this case I think its ok to go with simplest approach. Migration code looks ok to me, so may not worth it to spend time hacking something in ansible

shanyp commented:
yes, actually we have the  /release-notes label for it

LizardWizzard commented:
Would be cool to revive this patch

LizardWizzard commented:
As discussed moving to draft (for now)

problame commented:
As of many months ago, we're always attaching on startup, there's no more load path.
